### DISCUSSION TITLE
Milvus storage in AWS S3

### DISCUSSION THREAD

**Discussion by learning-stack** [milvus-io/milvus#26244] [Q&A and General discussion]
When I install Milvus and hook it up to AWS S3, I can only see the insert_log and stats_log folders inside AWS S3 bucket, but I thought it is also for blob storage (index files) as well.

Does anyone has any insight? Thank you!

**Comment by yhmo** [comment]
Index task is asynchronous. At the beginning, only the insert_log and stats_log folders are inside the bucket. Once an index task is finished, you will see index_files folder in the bucket.

**Reply by learning-stack** [reply]
Thank you, yhmo! I did try to import the documents to Milvus, and I can only see these 2 folders in AWS S3 created with some files. But I don't see the index_files folder at all. Do I need to do any steps further to get them in sync? I do see the content stored inside the volume mount though.

**Reply by yhmo** [reply]
Did you call `collection.create_index()` to create index?

**Reply by ndtreviv** [reply]
I'm having the same issue. I'm only seeing the following folders turn up in S3:

* `delta_log`
* `insert_log`
* `stats_log`

I'm using the Java SDK to create a collection and an index in that collection, and insert data and flush the collections. I can confirm that hashes are going into the index and I can search them, but when I shut the system down and start it all up again the data is all gone.

**Reply by ndtreviv** [reply]
I just started getting `index_files`. Can you please tell us how often the index files are written? I need to know if I update my index and then shut something down that the latest data is going to be represented.

**Reply by yhmo** [reply]
Index node begins to build index if a segment is flushed to be a sealed segment, and the segment row number exceeds this configuration:
```
indexCoord:
  minSegmentNumRowsToEnableIndex: 1024 # It's a threshold. When the segment num rows is less than this value, the segment will not be indexed
```
The time interval to check segments is 1 minute.

**Reply by ndtreviv** [reply]
OK. Let’s say I have 500 vectors that won’t be written to an index_file because it’s below this threshold.
If the system restarts am I expecting to lose that data? Is there any way I can force all segments to be written, say, before system shutdown?

EDIT:
I've just seen this:

> The flush() call will seal any remaining segments and send them for indexing. It is important to only call this at the end of an upsert session, as calling this too much will cause fragmented data that will need to be cleaned later on.

Despite flushing I wasn't seeing index files, but perhaps this was because they hadn't had time to be indexed before the system shut down?

**Reply by ndtreviv** [reply]
I think there's something else going on here. When writing data to S3, if no index_files are written then my collections don't persist.

Here's my test process:

1. start the system up (configured so that minio writes files to S3)
2. create a collection with an index
3. index some data (<1024 rows), flush the collection
4. GetCollectionStats and GetCollectionStatus to confirm the collection is there with vectors in it
5. Wait for 20 minutes - no index_files appear.
6. Shut the system down, start it up again
7. try to load collections, but they're not there, so it re-creates them.

I've attached two log files: milvus.log for before the shutdown (items 1-5), and milvus-after.log for after the restart (items 6 & 7).

[milvus-after.log](https://github.com/milvus-io/milvus/files/12521166/milvus-after.log)
[milvus.log](https://github.com/milvus-io/milvus/files/12521167/milvus.log)

**Reply by ndtreviv** [reply]
I'm going to open another discussion that better represents this issue.

**Reply by learning-stack** [reply]
Hi @ndtreviv, May I know how would you get the index_files folder? Are you using Milvus standalone? Thank you!

**Reply by ndtreviv** [reply]
@learning-stack I'm using milvus standalone. I've since figured out the reason for my missing data was that I didn't mount the etcd volume correctly, so the etcd data was being wiped out on restart. 
This brings me back to my original requirement: I was hoping to run a setup _not_ in k8s that stored all of its state in S3.
It seems that only some data is stored in S3, so I'll architect my way around it some other way.
